File: native-wrappers/android/datadogwrapper/src/main/kotlin/com/datadog/wrapper/DdSessionReplay.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.wrapper

import android.util.Log
import com.datadog.android.Datadog
import com.datadog.android.sessionreplay.SessionReplay
import com.datadog.android.sessionreplay.SessionReplayConfiguration
import com.datadog.android.sessionreplay.ImagePrivacy
import com.datadog.android.sessionreplay.TextAndInputPrivacy
import com.datadog.android.sessionreplay.TouchPrivacy

class DdSessionReplay {
    companion object {

        @JvmStatic
        fun mapTextAndInputPrivacy(level: String): TextAndInputPrivacy =
            when (level.lowercase()) {
                "mask_all_inputs" -> TextAndInputPrivacy.MASK_ALL_INPUTS
                "mask_sensitive_inputs" -> TextAndInputPrivacy.MASK_SENSITIVE_INPUTS
                else -> TextAndInputPrivacy.MASK_ALL
            }

        @JvmStatic
        fun mapImagePrivacy(level: String): ImagePrivacy =
            when (level.lowercase()) {
                "mask_none" -> ImagePrivacy.MASK_NONE
                else -> ImagePrivacy.MASK_ALL
            }

        @JvmStatic
        fun mapTouchPrivacy(level: String): TouchPrivacy =
            when (level.lowercase()) {
                "show" -> TouchPrivacy.SHOW
                else -> TouchPrivacy.HIDE
            }

        @JvmStatic
        fun enableSessionReplay(
            replaySampleRate: Double,
            textAndInputPrivacy: String,
            imagePrivacy: String,
            touchPrivacy: String,
            customEndpoint: String?
        ) {
            try {
                val builder = SessionReplayConfiguration.Builder(replaySampleRate.toFloat())
                    .setTextAndInputPrivacy(mapTextAndInputPrivacy(textAndInputPrivacy))
                    .setImagePrivacy(mapImagePrivacy(imagePrivacy))
                    .setTouchPrivacy(mapTouchPrivacy(touchPrivacy))

                if (!customEndpoint.isNullOrBlank()) {
                    builder.useCustomEndpoint(customEndpoint)
                }

                val config = builder.build()
                SessionReplay.enable(config, Datadog.getInstance())
            } catch (e: Exception) {
                Log.e("DatadogWrapper", "DdSessionReplay.enableSessionReplay failed", e)
            }
        }
    }
}

File: native-wrappers/ios/DatadogWrapper/Sources/DatadogWrapper/DdSessionReplay.swift
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2026-Present Datadog, Inc.
 */

import Foundation
import DatadogSessionReplay

@objc(DdSessionReplay)
public class DdSessionReplay: NSObject {

    // MARK: - Mapping helpers

    static func mapTextAndInputPrivacy(_ level: String) -> TextAndInputPrivacyLevel {
        switch level.lowercased() {
        case "mask_all_inputs": return .maskAllInputs
        case "mask_sensitive_inputs": return .maskSensitiveInputs
        default: return .maskAll
        }
    }

    static func mapImagePrivacy(_ level: String) -> ImagePrivacyLevel {
        switch level.lowercased() {
        case "mask_none": return .maskNone
        default: return .maskAll
        }
    }

    static func mapTouchPrivacy(_ level: String) -> TouchPrivacyLevel {
        switch level.lowercased() {
        case "show": return .show
        default: return .hide
        }
    }

    // MARK: - Enable

    @objc(enableSessionReplay:textAndInputPrivacy:imagePrivacy:touchPrivacy:customEndpoint:)
    public static func enableSessionReplay(
        replaySampleRate: Double,
        textAndInputPrivacy: String,
        imagePrivacy: String,
        touchPrivacy: String,
        customEndpoint: String?
    ) {
        var config = SessionReplay.Configuration(
            replaySampleRate: Float(replaySampleRate),
            textAndInputPrivacyLevel: mapTextAndInputPrivacy(textAndInputPrivacy),
            imagePrivacyLevel: mapImagePrivacy(imagePrivacy),
            touchPrivacyLevel: mapTouchPrivacy(touchPrivacy)
        )

        if let customEndpoint = customEndpoint,
           !customEndpoint.isEmpty,
           let url = URL(string: customEndpoint) {
            config.customEndpoint = url
        }

        SessionReplay.enable(with: config)
    }
}
